timestamp-client/src/test/java/com/palantir/timestamp/RequestBatchingTimestampServiceTest.java

```diff
@@ -44,60 +75,67 @@ public void delegatesInitializationCheck() {
     }
 
     @Test
-    public void testRateLimiting() throws Exception {
-        final long minRequestMillis = 100L;
-        final long testDurationMillis = 2000L;
-        final int numThreads = 3;
-
-        final StatsTrackingTimestampService rawTs = new StatsTrackingTimestampService(new InMemoryTimestampService());
-        final RequestBatchingTimestampService cachedTs = new RequestBatchingTimestampService(rawTs, minRequestMillis);
-        final AtomicLong timestampsGenerated = new AtomicLong(0);
-        final long startMillis = System.currentTimeMillis();
-
-        ExecutorService executor = PTExecutors.newCachedThreadPool();
-        try {
-            for (int i = 0; i < numThreads; ++i) {
-                executor.submit(() -> {
-                    while (System.currentTimeMillis() - startMillis < testDurationMillis) {
-                        cachedTs.getFreshTimestamp();
-                        timestampsGenerated.incrementAndGet();
-                    }
-                    return null;
-                });
-            }
-        } finally {
-            executor.shutdown();
-            executor.awaitTermination(1000, TimeUnit.SECONDS);
-        }
+    public void throwsIfAskForZeroTimestamps() {
+        assertThatThrownBy(() -> timestamp.getFreshTimestamps(0))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessage("Cannot ask for 0 timestamps");
+    }
 
-        assertEquals(0, rawTs.getFreshTimestampReqCount.get());
-        long approxFreshTimestampReqTotal = rawTs.getFreshTimestampsReqCount.get() * numThreads;
-        assertEquals(approxFreshTimestampReqTotal, timestampsGenerated.get(), approxFreshTimestampReqTotal);
+    @Test
+    public void coalescesRequestsTogether() {
+        assertThat(requestBatches(1, 2, 3)).containsExactly(single(1), range(2, 4), range(4, 7));
+        verify(unbatchedDelegate).getFreshTimestamps(6);
+        verifyNoMoreInteractions(unbatchedDelegate);
     }
 
-    private static class StatsTrackingTimestampService implements TimestampService {
-        AtomicLong getFreshTimestampReqCount = new AtomicLong(0);
-        AtomicLong getFreshTimestampsReqCount = new AtomicLong(0);
-        AtomicLong timestampsCount = new AtomicLong(0);
+    @Test
+    public void handsOutMaximumNumberOfTimestampsIfLimited() {
+        assertThat(requestBatches(1, 20_000, 3))
+                .containsExactly(single(1), range(10_001, 20_001), range(20_001, 20_004));
+        verify(unbatchedDelegate).getFreshTimestamps(20_004);
+        verify(unbatchedDelegate).getFreshTimestamps(20_003);
+        verify(unbatchedDelegate).getFreshTimestamps(3);
+        verifyNoMoreInteractions(unbatchedDelegate);
```

Comment:
Can do. This seems a little stronger than what our APIs guarantee, but I'm not sure how I would change it anyway.